java/monkey/src/main/java/root/parser/Parser.java

```diff
@@ -0,0 +1,112 @@
+import root.ast.statements.LetStatement;
+import root.ast.statements.ReturnStatement;
+import root.ast.statements.Statement;
+import root.lexer.Lexer;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Parser {
+
+    private final Lexer lexer;
+
+    private Token currentToken;
+
+    private Token peekToken;
+
+    public final List<String> errors = new ArrayList<>();
+
+    public Parser(Lexer lexer) {
+        this.lexer = lexer;
+
+        this.nextToken();
+        this.nextToken();
+    }
+
+    public Program parseProgram() {
+        Program program = new Program();
+
+        while (!curTokenIs(TokenType.EOF)) {
+            try {
+                Statement statement = parseStatement();
+                if (statement != null) {
+                    program.getStatements().add(statement);
+                }
+            } catch (ParserException pe) {
+                this.errors.add(pe.getMessage());
+            }
+            nextToken();
+        }
+
+        return program;
+    }
+
+    private void peekError(TokenType expected) throws ParserException {
+        throw new ParserException("Expected next token to be %s, got %s".formatted(expected.name(), this.currentToken.type().name()));
+    }
+
+    private void nextToken() {
+        this.currentToken = this.peekToken;
+        this.peekToken = this.lexer.nextToken();
+    }
+
+    private Statement parseStatement() throws ParserException {
+        return switch (this.currentToken.type()) {
+            case LET -> this.parseLetStatement();
+            case RETURN -> this.parseReturnStatement();
+            default -> null;
+        };
+    }
+
+    private Statement parseReturnStatement() {
+        ReturnStatement returnStatement = new ReturnStatement(this.currentToken);
+
+        this.nextToken();
+
+        // TODO: We're skipping the expressions until we encounter a semicolon or EOF
+        while (!this.curTokenIs(TokenType.SEMI) && !this.curTokenIs(TokenType.EOF)) {
+            this.nextToken();
+        }
+
+        return returnStatement;
+    }
+
+    private Statement parseLetStatement() throws ParserException {
+        LetStatement letStatement = new LetStatement(this.currentToken);
+
+        this.expectPeek(TokenType.IDENT);
+
+        letStatement.setName(new IdentiferExpression(this.currentToken, this.currentToken.literal()));
+
+        this.expectPeek(TokenType.ASSIGN);
+
+        // TODO: We're skipping the expressions until we encounter a semicolon or EOF
+        while (!this.curTokenIs(TokenType.SEMI) && !this.curTokenIs(TokenType.EOF)) {
+            this.nextToken();
+        }
+
+        return letStatement;
+    }
+
+    private boolean curTokenIs(TokenType type) {
+        return this.currentToken.type() == type;
+    }
+
+    private boolean peekTokenIs(TokenType type) {
+        return this.peekToken.type() == type;
+    }
+
+    private void expectPeek(TokenType type) throws ParserException {
```

Comment:
Given the comment about `ExceptionStatement` and such, we'd probably swap this to eventually be `private boolean expectPeek(TokenType type)`, so then we can begin doing something like this.
```java
switch (foo) {
     case LetStatement ls when expectPeek(TokenType.INDENT) -> //do something
     case LetStatement ls when !expectPeek(TokenType.INDENT) -> // do something else
}
```
Once again, just thinking about future stuff, so we can hold it off till we hit the future.

Reply:
It could be done but I don't see a reason to use switch here other than having a single return statement. We are not matching anything, foo here is always a LetStatement, if statements should be enough. I personally don't see having multiple return statements as that big of a problem in simple functions where the early returns are just error cases, I can see the point where multiple returns return valid outputs, but not for error cases.